x-pack/plugins/security_solution/public/common/utils/global_query_string/index.ts

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import * as H from 'history';
+import { parse, ParsedQuery, stringify } from 'query-string';
+import { useCallback, useEffect, useMemo } from 'react';
+
+import { url } from '@kbn/kibana-utils-plugin/public';
+import { isEmpty, pickBy } from 'lodash/fp';
+import { useHistory } from 'react-router-dom';
+import { useDispatch } from 'react-redux';
+import {
+  decodeRisonUrlState,
+  encodeRisonUrlState,
+  getParamFromQueryString,
+  getQueryStringFromLocation,
+} from '../../components/url_state/helpers';
+import { useShallowEqualSelector } from '../../hooks/use_selector';
+import { globalUrlParamActions, globalUrlParamSelectors } from '../../store/global_url_param';
+import { useRouteSpy } from '../route/use_route_spy';
+import { getLinkInfo } from '../../links';
+
+/**
+ * Adds urlParamKey and the initial value to redux store.
```

Comment:
Is the only way you want url params to be de-registered is when this effect unmounts or should devs be free to call `deregisterUrlParam` from within a page level component?

Reply:
The `useInitializeUrlParam` hook covers registering and de-registering the URL param. So, devs should not dispatch any `globalUrlParamActions`. 

I designed the `useInitializeUrlParam` hook to be called from top-level components that should not be unmounted during page changes. But in theory, it would work if called from a page level component. But it would be slower because it would have to read and decode the data from the `window.location` on every mount.

Reply:
Makes sense. I wonder then if it just makes more sense to have an enhanced route component or HOC that shows a clear pattern for keeping it at the top level, rather than relying on devs to remember to add it to the top level component. 

```
const SecurityRoute = ({ component: Component, urlParamKey, onInitialize, ...rest }) =>  {
 useInitializeUrlParam(urlParamKey, onInitialize);
// Any other top level things that need to be done for this route
  return (
    <Route
      {...rest}
      component={Component}
    />
  );
}

```

Reply:
**Documenting our discussion:** We are now using global query string hooks on global components that are initialized outside the router. But it is a good idea to simplify the usage of global query strings so we can make it foolproof.